### PR TITLE
Bring @microstates/ember up to date with Microstates.js library

### PR DESCRIPTION
**Pull request by taras**
It's been more than 2 years since Charles' last commit to this repository. Since then, most of the work was done in [cowboyd/microstates.js](http://github.com/cowboyd/microstates.js) and now [microstates/microstates.js](http://github.com/microstates/microstates.js) repository. 

A lot has happened in Microstates world since `ember-microstates` was last worked on. This PR aims to bring the Ember integration with Microstates with the latest version of Microstates. There are some significant breaking changes to maintain consistency with the main Microstates library.

In spirit, the Ember add-on remains primarily a way to make templates more expressive and eliminate the need for writing actions to perform common state transitions. In `ember-microstates` this is only possible with types that were provided by `ember-microstates` namely String, Boolean, Number, List, Object,  Select and MultiSelect. It was not possible to create custom types because types in `ember-microstates` are not composable. 

The new version of Microstates allows to compose Microstate types into complex data structures. Microstates will provide the transitions that you can perform on Microstates instances created from the Microstates types. `@microstates/ember` will ensure that the result of these transitions will be reflected in the template. You can learn about types and type composition in [the Microstates README](https://github.com/microstates/microstates.js#types-and-type-composition).

## New API

- helpers:
  - `(state value)` - creates a microstate from a value and wraps it in a store
  - `(type name)` - retrieves type of a given name
  - `(value-of microstate)` - gets value from microstate
- computed property macros
  - `state(microstate)`

When using primitive types in templates, the developer can do this.

```hbs
{{#let (state 42) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

When the developer needs to create a microstate, they can just do

```hbs
{{#let (state (type "number") 10) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

You can use custom types by putting JS classes into files in `/types/` directory.

```js
// file in /app/types/person.js
export default class Person {
  name = String;
  age = Number;
}
```

```hbs
{{#let (state (type "person") initial) as |taras|}}
  {{taras.name.state}} - {{taras.age.state}}
{{/let}}
```

In JavaScript, the developer can do this,

```js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { valueOf, state, create, from } from '@microstates/ember';

export default Component.extend({
  // if the type is not necessary: `state([10, 20, 30])`
  numbers: state([Number], [10, 20, 30]),

  value: computed('numbers', function() {
    return valueOf(this.get('numbers'));
  }),
  
  layout: hbs`
  {{#each numbers as |number|}}
    <ul>
      <li><button {{action numbers.remove number}}>{{number.state}}</button></li>
    </ul>
  {{/each}}
  `
})
```

### Built-in Types

`ember-microstates` re-exports types from `microstates` package. This is necessary because we're exporting modules from the CJS build instead of the ES build that `ember-auto-import` uses by default. The CJS build does not transpile the classes which avoids the need for adding IE11 as a target in development.

| Type     | JavaScript | `(type) helper` |
|----------|------------|-----------------|
| Any  | `import {Any} from '@microstates/ember'` | `(type 'any')` |
| Boolean  | `import {BooleanType} from '@microstates/ember'` | `(type 'boolean')` |
| Number  | `import {NumberType} from '@microstates/ember'` | `(type 'number')` |
| String  | `import {StringType} from '@microstates/ember'` | `(type 'string')` |
| Object  | `import {ObjectType} from '@microstates/ember'` | `(type 'object')`  |
| Array  | `import {ArrayType} from '@microstates/ember'` | `(type 'array')` |

## Serializing Microstates

You will likely need to be able to capture the value of the microstate to persist it. You can do this with `(value-of microstate)` and `valueOf(microstate)` function.

```hbs
{{#let (state 42) as |$|}}
  <button {{action $.increment}}>Increment</button>
  <button {{action (mut value) (value-of $)}}>Save</button>
{{/let}}
```

```js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { valueOf, state } from '@microstates/ember';

export default Component.extend({
  numbers: state([10, 20, 30]),

  value: computed('numbers', function() {
    return valueOf(this.get('numbers'));
  }),
  
  layout: hbs`
  {{#each numbers as |number|}}
    <ul>
      <li><button {{action numbers.delete number}}>{{number.state}}</button></li>
    </ul>
  {{/each}}
  `
})
```

## TODO

- [x] Wait for https://github.com/microstates/microstates.js/pull/276 to be released
- [x] Add integration tests for all helpers
- [x] Move to `@microstates/ember`

### CODE REVIEW

**Comment by taras** [comment]
@alexlafroscia @knownasilya what do you guys think?

**Comment by knownasilya** [comment]
Looks good to me!

**Comment by alexlafroscia** [comment]
I like the idea too, but am curious about this `import` helper -- I feel like it would be a little more "natural" to have something like a `type` helper that goes through the resolver to look up the module.

With the same `types/Person` file, it might look something like this:

```hbs
{{#let (use-type (type 'person')) as |person|}}
    {{person.name.state}} is {{person.age.state}}
{{/let}}
```

I think that the `type` helper could look something like this:

```javascript
import Helper from '@ember/component/helper';
import { getOwner } from '@ember/application';

export default Helper.extend({
  compute([name]) {
    return getOwner(this).lookup(`type:${name}`);
  }
});
```

This is a little more "terse" in that you don't need the `my-app/types` in the import name

**Comment by knownasilya** [comment]
I'm sure `ember-import` will come in handy in other scenarios, like types that are shared on npm and not built as ember addons.

**Comment by alexlafroscia** [comment]
I definitely don't mean to disparage the idea of `ember-import` as a whole -- I just think that a more "tailored" helper for this use-case that's specific to looking up a type would be a nicer API, and as a bonus avoid using the global `require` function which, while not exactly "private" API, might be better to avoid.

**Comment by knownasilya** [comment]
Bam! @taras is on a roll!

**Comment by taras** [comment]
I pushed latest version of `ember-microstates` to this branch and updated the README. 

cc @sivakumar-kailasam

**Comment by taras** [comment]
@alexlafroscia it's a great idea to eliminate the `useState` and `useType` CPMs. It got me thinking that the API surface might be bigger than necessary in general. 

The API surface will be
- helpers:
  - `(use microstate)` - wraps Microstate in a store
  - `(create Type value)` - creates a microstate from type value
  - `(from value)` - creates microstate from value
  - `(value-of microstate)` - gets value from microstate
- computed property macros
  - `use(microstate)`

When using primitive types in templates, the developer can do this.

```hbs
{{#let (use (from 42)) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

When the developer needs to create a microstate, they can just do

```hbs
{{#let (use (create (type "person") (hash name="Taras"))) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

In JavaScript, the developer can do this,

```js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { valueOf, use, create, from } from 'ember-microstates';

export default Component.extend({
  // if the type is not necessary: `from([10, 20, 30])`
  numbers: use(create([Number], [10, 20, 30])),

  value: computed('numbers', function() {
    return valueOf(this.get('numbers'));
  }),
  
  layout: hbs`
  {{#each numbers as |number|}}
    <ul>
      <li><button {{action numbers.delete number}}>{{number.state}}</button></li>
    </ul>
  {{/each}}
  `
})
```

If Ember wasn't so "opinionated" the only API that would be necessary would be `use` helper and `macro`, but because templates can't use JavaScript functions we have to define `create`, `from` and `value-of` helpers. In JavaScript, they can just use functions exports from microstates.

This will be a huge simplification and a much better alignment.

Thoughts?

**Comment by knownasilya** [comment]
I like that direction as well, feels very composable 👍

**Comment by brandynbennett** [comment]
Will the `use` helper conflict with the `use` helper coming out in the new Module Unification API?

**Comment by taras** [comment]
@brandynbennett yes it will. Thank you for bringing it to my attention.

**Comment by taras** [comment]
Ok, since we can't use `use` helper, what about just using `state`.

* helpers
  * `(state valueOrMicrostate)` - takes a value, which it converts to a microstate with `Microstates.from`, or a microstate and wraps it in a `Store`
 * `(create Type value)` - creates a Microstate from a type
 * `(type name)` - retrieves the type
* macros
 * `state` - takes a value, which it converts to a microstate with `Microstates.from`, or a microstate and wraps it in a `Store`

```hbs
{{#let (state 42) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

```hbs
{{#let (state (type "person") initial) as |number|}}
  <button {{action number.increment}}>Increment ({{number.state}})</button>
{{/let}}
```

```js
import Component from '@ember/component';
import { computed } from '@ember/object';
import { valueOf, state, create } from 'ember-microstates';

export default Component.extend({
  // if the type is not necessary: `from([10, 20, 30])`
  numbers: state(create([Number], [10, 20, 30])),

  value: computed('numbers', function() {
    return valueOf(this.get('numbers'));
  }),
  
  layout: hbs`
  {{#each numbers as |number|}}
    <ul>
      <li><button {{action numbers.remove number}}>{{number.state}}</button></li>
    </ul>
  {{/each}}
  `
})
```

**Comment by brandynbennett** [comment]
Will you be keeping the `value-of` helper? I think that one would still be useful for getting objects/arrays

**Comment by taras** [comment]
> Will you be keeping the `value-of` helper? I think that one would still be useful for getting objects/arrays

Yes, I just added it to the description. I'm going to finalize this now.

**Comment by taras** [comment]
I just pushed TodoMVC example. Here is what entire TODOMVC app looks like in Ember - without writing JavaScript. It's using [TodoMVC type](https://github.com/microstates/todomvc/blob/4f5d1e770f55b0b4e8b65500f29f6552cc413378/todomvc.js) from `@microstates/todomvc`. 

```hbs
   {{#let (state (type "todomvc") initialTodomvc) as |todomvc|}}
        <section class="todoapp">
          <form class="header" {{action todomvc.insertNewTodo on="submit"}}>
            <h1>todos</h1>
            <input
              class="new-todo"
              placeholder="What needs to be done?"
              value={{todomvc.newTodo.state}}
              onchange={{action todomvc.newTodo.set value="target.value"}}
              autofocus
            />
          </form>
          <section class="main">
            <input class="toggle-all" type="checkbox" checked={{todomvc.isAllComplete}} />
            <label for="toggle-all" {{action todomvc.toggleAll}}>
              Mark all as complete
            </label>
            <ul class="todo-list">
              {{#each todomvc.filtered as |todo|}}
                <li class={{if todo.completed.state "completed"}}>
                  {{#if todo.editing.state}}
                    <input 
                      class="edit" 
                      value={{todo.text.state}} 
                      onchange={{action todo.text.set value="target.value"}}
                      onblur={{action todo.save}}
                    />
                  {{else}}
                    <div class="view">
                      <input 
                        class="toggle" 
                        type="checkbox" 
                        checked={{todo.completed.state}} 
                        onclick={{action todo.completed.toggle}} 
                      />
                      <label {{action todo.edit on="doubleClick"}}>
                        {{todo.text.state}}
                      </label>
                      <button class="destroy" {{action todomvc.todos.remove todo}} />
                    </div>
                  {{/if}}
                </li>
              {{/each}}
            </ul>
          </section>
          <footer class="footer">
            <span class="todo-count">
              <strong>
                {{todomvc.active.length}}
              </strong>
              {{if (eq todomvc.active.length 1) 'item' 'items'}} left
            </span>
            <ul class="filters">
              {{#each todomvc.filters as |filter|}}
                <li>
                  <button
                    class={{if filter.selected "selected"}}
                    {{action filter.select}}
                  >
                    {{filter.label}}
                  </button>
                </li>
              {{/each}}
            </ul>
            {{#if todomvc.hasCompleted}}
              <button class="clear-completed" {{action todomvc.clearCompleted}}>
                Clear completed
              </button>
            {{/if}}
          </footer>
        </section>
      {{/let}}
``` 

*Note:* I discovered and fixed a bug in Microstates. This latest PR requires fixed version of Microstates. You can see the [PR here](https://github.com/microstates/microstates.js/pull/284).

**Comment by knownasilya** [comment]
This PR seems to be getting quite heavy 😸 maybe time to merge and open smaller PRs to fix subsequent issues?

**Comment by taras** [comment]
It's getting meaty AF. I'm waiting for https://github.com/microstates/microstates.js/pull/284#issuecomment-441796876 to be merged so I can merge this PR with working demos. Right now, the demos work only with `yarn link`. I suspect today or tomorrow we'll have it merged.

Then you can feast your eyes on the new demo app :)

**Comment by taras** [comment]
Ok, this is ready to go. Here is the updated demo site https://5bfcbca19a063f20024980b5--ember-microstates.netlify.com

File: package.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
-  "name": "ember-microstates",
+  "name": "@microstates/ember",
   "version": "0.3.0",
   "description": "Data Down, Actions Up: at the molecular level.",
   "keywords": [
     "ember-addon"
   ],
-  "repository": "https://github.com/cowboyd/ember-microstates.git",
+  "repository": "https://github.com/cowboyd/@microstates/ember.git",
```
Comment (sivakumar-kailasam) [inline package.json]:
```suggestion
  "repository": "https://github.com/microstates/ember.git",
```

File: package.json [hunk 2]
```diff
@@ -1,11 +1,11 @@
 {
-  "name": "ember-microstates",
+  "name": "@microstates/ember",
   "version": "0.3.0",
   "description": "Data Down, Actions Up: at the molecular level.",
   "keywords": [
     "ember-addon"
```
Comment (sivakumar-kailasam) [inline package.json]:
```suggestion
    "ember-addon",
    "microstates"
```

**Pull request merged**